airbyte-integrations/connectors/source-posthog/source_posthog/streams.py

```diff
@@ -0,0 +1,312 @@
+#
+
+
+import math
+import urllib.parse
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class PosthogStream(HttpStream, ABC):
+    url_base = "https://app.posthog.com/api/"
+    primary_key = "id"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if "next" in resp_json and resp_json["next"]:
+            next_query_string = urllib.parse.urlsplit(resp_json["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        mandatory_headers = {}
+        mandatory_headers["Content-Type"] = "application/json"
+        mandatory_headers["User-Agent"] = "posthog-python/1.4.0"
+        return mandatory_headers
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        yield from response_json[self.data_field]
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        params = {}
+        if next_page_token:
+            params.update(next_page_token)
+        return params
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """the responce entry that contains useful data"""
+        pass
+
+
+class IncrementalPosthogStream(PosthogStream, ABC):
+    state_checkpoint_interval = math.inf
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # params for desendant sorted responses
+        self.reversed_pagination = {"is_completed": False, "latest_response_state": {}, "is_new_page_available": False, "upgrade_to": {}}
+
+    @property
+    @abstractmethod
+    def cursor_field(self) -> str:
+        """
+        Defining a cursor field indicates that a stream is incremental, so any incremental stream must extend this class
+        and define a cursor field.
+        """
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if self.reversed_pagination["is_completed"]:
+            return {}
+        if "next" in resp_json and resp_json["next"]:
+            next_query_string = urllib.parse.urlsplit(resp_json["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
Since this function also returns a tuple, could you please specify this fact in return type? Because right now it contains only `Mapping[str, Any]`.